File: C++/tree/segtree.cpp
template<class T> class segtree{
private:
    int n, size;
    vector<T> seg;
    T e;
    function<T(T, T)> op;
public:
    segtree(const vector<T>& A, T id, function<T(T, T)> op) : e(id), op(op){
        n = A.size();
        size = 1;
        while (size < n) size <<= 1;
        seg.assign(2*size, e);
        for (int i = 0; i < n; i++) seg[size+i] = A[i];
        for (int i = size-1; i > 0; i--) seg[i] = op(seg[i<<1], seg[i<<1|1]);
    }

    void update(int i, T val){
        i += size;
        seg[i] = val;
        while (i >>= 1) seg[i] = op(seg[i<<1], seg[i<<1|1]);
    }

    T query(int l, int r){
        T L = e, R = e;
        for (l += size, r += size; l < r; l >>= 1, r >>= 1){
            if (l&1) L = op(L, seg[l++]);
            if (r&1) R = op(seg[--r], R);
        }
        return op(L, R);
    }

    T get(int i){
        return seg[size+i];
    }

    void add(int i, T val){
        update(i, get(i)+val);
    }

    template<typename F> int max_right(int l, F f){
        if (l == n) return n;
        l += size;
        T sm = e;
        do{
            while ((l & 1) == 0) l >>= 1;
            if (!f(op(sm, seg[l]))){
                while (l < size){
                    l <<= 1;
                    if (f(op(sm, seg[l]))){
                        sm = op(sm, seg[l]);
                        l++;
                    }
                }
                return l-size;
            }
            sm = op(sm, seg[l]);
            l++;
        }while((l&-l) != l);
        return n;
    }

    template<typename F> int min_left(int r, F f){
        if (r == 0) return 0;
        r += size;
        T sm = e;
        do{
            r--;
            while (r > 1 && (r&1)) r >>= 1;
            if (!f(op(seg[r], sm))){
                while (r < size){
                    r = r<<1|1;
                    if (f(op(seg[r], sm))){
                        sm = op(seg[r], sm);
                        r--;
                    }
                }
                return r+1-size;
            }
            sm = op(seg[r], sm);
        }while((r&-r) != r);
        return 0;
    }
};

File: py/segtree.py
class segtree:
    def __init__(self, arg, op, e):
        self.op = op
        self.e = e
        if isinstance(arg, list):
            A = arg
            self.n = len(A)
            self.size = 1
            while self.size < self.n:
                self.size <<= 1
            self.seg = [e]*(2*self.size)
            for i in range(self.n):
                self.seg[self.size+i] = A[i]
            for i in range(self.size-1, 0, -1):
                self.seg[i] = op(self.seg[i<<1], self.seg[i<<1|1])
        else:
            self.n = arg
            self.size = 1
            while self.size < self.n:
                self.size <<= 1
            self.seg = [e]*(2*self.size)
            for i in range(self.size-1, 0, -1):
                self.seg[i] = op(self.seg[i<<1], self.seg[i<<1|1])

    def set(self, i, val):
        i += self.size
        self.seg[i] = val
        i >>= 1
        while i:
            self.seg[i] = self.op(self.seg[i<<1], self.seg[i<<1|1])
            i >>= 1

    def all_prod(self):
        return self.seg[1]

    def prod(self, l, r):
        L = self.e
        R = self.e
        l += self.size
        r += self.size
        while l < r:
            if l&1:
                L = self.op(L, self.seg[l])
                l += 1
            if r&1:
                r -= 1
                R = self.op(self.seg[r], R)
            l >>= 1
            r >>= 1
        return self.op(L, R)

    def get(self, i):
        return self.seg[self.size+i]

    def __getitem__(self, i):
        return self.seg[self.size+i]

    def add(self, i, val):
        self.set(i, self.get(i)+val)

    def max_right(self, l, f):
        if l == self.n:
            return self.n
        l += self.size
        sm = self.e
        while True:
            while (l&1) == 0:
                l >>= 1
            if not f(self.op(sm, self.seg[l])):
                while l < self.size:
                    l <<= 1
                    if f(self.op(sm, self.seg[l])):
                        sm = self.op(sm, self.seg[l])
                        l += 1
                return l-self.size
            sm = self.op(sm, self.seg[l])
            l += 1
            if (l&-l) == l:
                break
        return self.n

    def min_left(self, r, f):
        if r == 0:
            return 0
        r += self.size
        sm = self.e
        while True:
            r -= 1
            while r > 1 and (r&1):
                r >>= 1
            if not f(self.op(self.seg[r], sm)):
                while r < self.size:
                    r = r<<1|1
                    if f(self.op(self.seg[r], sm)):
                        sm = self.op(self.seg[r], sm)
                        r -= 1
                return r+1-self.size
            sm = self.op(self.seg[r], sm)
            if (r&-r) == r:
                break
            return 0
